### PR TITLE
Implement reads in KVS

### PR DESCRIPTION
**Pull request by justinj**
First two commits are courtesy of Mingwei to make some lattice stuff work.

### CODE REVIEW

File: kvs/src/main.rs [hunk 1]
```diff
@@ -152,16 +172,17 @@ where
                     }
                 });
 
-                let (reads_send, reads_recv) =
-                    hf.add_channel_input::<_, Option<K>, VecHandoff<K>>("reads");
-
-                let _ = reads_send;
+                // let (reads_send, reads_recv) =
+                //     hf.add_channel_input::<_, Option<(K, futures::channel::oneshot::Sender<V>)>,
+                //     VecHandoff<(K, futures::channel::oneshot::Sender<V>)>>("reads");
```
Comment (MingweiSamuel) [inline kvs/src/main.rs:177]: Cleanup since this moved to L184

File: kvs/src/main.rs [hunk 2]
```diff
@@ -216,27 +238,45 @@ where
...
-                        println!("got batch: {:?} {:?} {:?}", id, epoch, batch);
-                        // TODO(justin): this doesn't do the clocks. I
-                        // think we need to do some weird join against a
-                        // singleton that I don't really understand yet.
-                        batch.into_iter()
-                    })).pull_to_push().for_each(|x| {
-                        println!("read: {:?}", x)
-                    })
+                let my_id = id;
+                hf.add_subgraph("read_handler", reads_recv.flatten().map(|(k, ch)| (k, ch)).half_hash_join::<_, _, _, ClockedDataRepr<V>, ClockedUpdateRepr<V>>(
```
Comment (MingweiSamuel) [inline kvs/src/main.rs:264]: Is rustfmt failing here? Can we space it out a bit more

**Review by MingweiSamuel — APPROVED** [review]
Yay

**Comment by MingweiSamuel** [comment]
And you can ignore the build

File: kvs/src/main.rs (hunk 1 above)
Comment (justinj) [reply kvs/src/main.rs:177]: Oops, good catch.

File: kvs/src/main.rs (hunk 2 above)
Comment (justinj) [reply kvs/src/main.rs:264]: Yeah I'm not sure why rustfmt does not seem to be working here. I will have to investigate at some point.

**Pull request merged**